### PR TITLE
Add and remove layers dynamically

### PR DESCRIPTION
**Pull request by jfirebaugh**
This PR implements the first bit of the core [runtime styling API](https://github.com/mapbox/mapbox-gl-native/issues/837): the ability to dynamically add and remove layers.

As appropriate for a C++ API, the layers API is _strongly typed_: there are subclasses for each layer type, with correctly-typed accessors for each style property. This results in a large API surface area. Fortunately, this is automated, by generating the API – and the regular portion of the implementation – from the style specification.

The layers API makes a distinction between public API and internal implementation using [the `Impl` idiom](https://github.com/mapbox/mapbox-gl-native/issues/3254) seen elsewhere in the codebase. Here, it takes the form of two parallel class hierarchies:

* `Layer` and its subclasses form the public API.
* `Layer::Impl` and its subclasses form the internal API.

As well as forming the boundary between public and internal, these two class hierarchies form the boundary between generated code and handwritten code. Except for `CustomLayer` and `CustomLayer::Impl`:

* `Layer` subclasses are entirely generated. (`Layer` itself is handwritten.)
* `Layer::Impl` and its subclasses are entirely handwritten.

--------

There is still work to be done before this is ready to be exposed via SDK bindings:

* [x] The types relating to property values need to be cleaned up. Conceptually, a property value is a variant with three possibilities: not set (equivalent to a default value), constant value, or function value. The property API needs to reflect this. Currently every value is treated as a function value.
* [x] The namespacing needs to be rationalized. Everything style-related should live in the `mbgl::style` namespace and `mbgl/style` directory. (Beneath that directory there can be a `layer` directory for organization.)
* [x] Accessors for the layer filter and zoom range need to be added.
* [ ] There needs to be a layer getter. I'm leaning towards `map.getStyle().getLayer()` as the API. `Map::{add,remove}Layer` would also move to `Style`.

### CODE REVIEW

File: include/mbgl/layer/background_layer.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    Function<std::string> getBackgroundPattern() const;
+    void setBackgroundPattern(Function<std::string>);
+
+    Function<float> getBackgroundOpacity() const;
+    void setBackgroundOpacity(Function<float>);
+
+    // Private implementation
+
+    std::unique_ptr<Layer> clone() const final;
```
Comment (kkaefer) [inline include/mbgl/layer/background_layer.hpp]: Any way we can move this to `Impl`?

File: scripts/generate-style-code.js [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    return property.default;
+  }
+}
+
+const layerHpp = ejs.compile(`<%
+  const type = locals.type;
+  const layoutProperties = locals.layoutProperties;
+  const paintProperties = locals.paintProperties;
+-%>
+// This file is generated. Do not edit.
```
Comment (kkaefer) [inline scripts/generate-style-code.js]: Can we add hints to how to regenerate the files in there?

File: include/mbgl/layer/background_layer.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+// This file is generated. Do not edit.
```
Comment (kkaefer) [inline include/mbgl/layer/background_layer.hpp]: Can we generate these files during the build, like we we with shader files?

**Comment by kkaefer** [comment]
Looks good! As for the generated files, it's probably useful to have them pre-generated to be able to look at them. We could still add the generation script to the build system, and check whether files were modified and fail if that is the case.

**Comment by 1ec5** [comment]
> We could still add the generation script to the build system, and check whether files were modified and fail if that is the case.

How about running the generation script as part of the build, but having the script output to the checkout, as it does now?

**Comment by jfirebaugh** [comment]
For now I'm going to go with running the generator manually when you make a change. I looked at including it as part of the build step but outputting to the src directory (which I want to do) didn't fit in cleanly with our use of gyp.

**Comment by jfirebaugh** [comment]
The last commit here involves a lot of renaming and re-namespacing. Would be great to get a second opinion on my approach:

* Everything "style-related" (roughly: implementing a concept from the style specification) lives in the `style` subdirectory and `mbgl::style` namespace.
* Inside the `mbgl::style` namespace, don't prefix class names with `Style` (except for the `Style` itself).
* Outside the `mbgl::style`, use forward declarations in headers where available.
* Use `using namespace style` in cpp files (not in headers).

**Pull request merged**